src/lib/protocol_version/protocol_version.ml

```diff
@@ -71,6 +71,104 @@ module Make_str (A : Wire_types.Concrete) = struct
      negative numbers
   *)
   let is_valid t = t.transaction >= 1 && t.network >= 1 && t.patch >= 0
+
+  let older_than_current t =
+    transaction t < transaction current
+    || (transaction t = transaction current && network t < network current)
+    || transaction t = transaction current
+       && network t = network current
+       && patch t < patch current
+
+  let gen =
+    let open Quickcheck.Generator.Let_syntax in
+    let%map transaction = Quickcheck.Generator.small_non_negative_int
+    and network = Quickcheck.Generator.small_non_negative_int
+    and patch = Quickcheck.Generator.small_non_negative_int in
+    { transaction = transaction + 1; network = network + 1; patch }
+
+  let deriver obj =
+    let open Fields_derivers_zkapps.Derivers in
+    let ( !. ) = ( !. ) ~t_fields_annots in
+    Fields.make_creator obj ~transaction:!.int ~network:!.int ~patch:!.int
+    |> finish "ProtocolVersion" ~t_toplevel_annots
+end
+
+module T = Wire_types.Make (Make_sig) (Make_str)
+include T
+
+module N = Mina_numbers.Nat.Make32 ()
+
+type var =
+  { transaction : N.Checked.t; network : N.Checked.t; patch : N.Checked.t }
+
+module Checked = struct
+  let to_input { transaction; network; patch } =
+    Random_oracle.Input.Chunked.(
+      append
+        (append (N.Checked.to_input transaction) (N.Checked.to_input network))
+        (N.Checked.to_input patch))
+
+  let if_ cond ~(then_ : var) ~(else_ : var) =
+    let open Snark_params.Tick in
+    let%map transaction =
+      N.Checked.if_ cond ~then_:then_.transaction ~else_:else_.transaction
+    and network = N.Checked.if_ cond ~then_:then_.network ~else_:else_.network
+    and patch = N.Checked.if_ cond ~then_:then_.patch ~else_:else_.patch in
+    { transaction; network; patch }
+
+  let constant (t : t) =
+    { transaction = N.Checked.constant (N.of_int (transaction t))
+    ; network = N.Checked.constant (N.of_int (network t))
+    ; patch = N.Checked.constant (N.of_int (patch t))
+    }
+
+  let current = constant current
+
+  let equal_to_current t =
```

Comment:
The RFC only mentions the `transaction` version, not the `network` and `patch` versions, when considering vk permissions. Is this function doing too much?

Reply:
Let me ask nathan, maybe that's too much.